### PR TITLE
uvf metadata (WiP).

### PR DESCRIPTION
**Pull request by chenkins**

### CODE REVIEW

File: frontend/src/common/crypto.ts [hunk 1]
```diff
@@ -141,7 +163,8 @@ export class VaultKeys {
         true,
         ['verify']
       );
-      return [new VaultKeys(await masterkey), { privateKey: await privKey, publicKey: await pubKey }];
+      // TODO https://github.com/encryption-alliance/unified-vault-format/pull/19 upstream legacy integration for uvf?
```
Comment (chenkins) [inline frontend/src/common/crypto.ts:166]: @overheadhunter upstream legacy integration for uvf?

File: frontend/src/common/crypto.ts [hunk 2]
```diff
@@ -174,7 +197,8 @@ export class VaultKeys {
       true,
       ['sign']
     );
-    return new VaultKeys(await key);
+    // TODO https://github.com/encryption-alliance/unified-vault-format/pull/19 upstream legacy integration for uvf?
```
Comment (chenkins) [inline frontend/src/common/crypto.ts:200]: @overheadhunter upstream legacy integration for uvf?

File: frontend/src/common/crypto.ts [hunk 3]
```diff
@@ -449,3 +475,116 @@ export async function getFingerprint(key: string | undefined) {
...
+
+  /**
+   * Creates new vault metadata with a new file key and name key
+   * @returns new vault metadata
+   */
+  public static async create(automaticAccessGrant: VaultMetadataJWEAutomaticAccessGrantDto): Promise<VaultMetadata> {
+    const fileKey = crypto.subtle.generateKey(
+      VaultMetadata.RAWKEY_KEY_DESIGNATION,
+      true,
+      // TODO https://github.com/encryption-alliance/unified-vault-format/pull/19 is this correct?
```
Comment (chenkins) [inline frontend/src/common/crypto.ts:507]: @overheadhunter review

File: frontend/src/common/crypto.ts [hunk 4]
```diff
@@ -449,3 +475,116 @@ export async function getFingerprint(key: string | undefined) {
...
+      ['sign']
+    );
+    const nameKey = crypto.subtle.generateKey(
+      VaultMetadata.RAWKEY_KEY_DESIGNATION,
+      true,
+      // TODO https://github.com/encryption-alliance/unified-vault-format/pull/19 is this correct?
```
Comment (chenkins) [inline frontend/src/common/crypto.ts:513]: @overheadhunter review

File: frontend/src/common/crypto.ts [hunk 5]
```diff
@@ -449,3 +475,116 @@ export async function getFingerprint(key: string | undefined) {
...
+   * @param jwe JWE containing the vault key
+   * @param masterKey the vault masterKey
+   * @returns vault metadata
+   */
+  public static async decryptWithMasterKey(jwe: string, masterKey: CryptoKey): Promise<VaultMetadata> {
+      const payload = await JWEParser.parse(jwe).decryptA256kw(masterKey);
+      const keys: Record<string,string> = payload['keys'];
+      const keysImported: Record<string,CryptoKey> = payload['keys'];
+      for (const k in keys) {
+        // TODO https://github.com/encryption-alliance/unified-vault-format/pull/19 is this correct?
```
Comment (chenkins) [inline frontend/src/common/crypto.ts:535]: @overheadhunter review

File: frontend/src/common/crypto.ts [hunk 6]
```diff
@@ -449,3 +475,116 @@ export async function getFingerprint(key: string | undefined) {
...
+    for (const k in this.keys) {
+      keysExported[k] = base64.stringify(new Uint8Array(await crypto.subtle.exportKey('raw', this.keys[k])));
+    }
+    const payload = {
+        fileFormat: "AES-256-GCM-32k",
+        nameFormat: "AES-256-SIV",
+        keys: keysExported,
+        latestFileKey: this.latestFileKey,
+        nameKey: this.nameKey,
+        // TODO https://github.com/encryption-alliance/unified-vault-format/pull/21 finalize kdf
```
Comment (chenkins) [inline frontend/src/common/crypto.ts:564]: @overheadhunter review

File: frontend/src/common/crypto.ts [hunk 7]
```diff
@@ -449,3 +475,116 @@ export async function getFingerprint(key: string | undefined) {
...
+        kdf: "1STEP-HMAC-SHA512",
+        'org.cryptomator.automaticAccessGrant': this.automaticAccessGrant
+    }
+    return JWEBuilder.a256kw(masterKey).encrypt(payload);
+  }
+
+  public async hashDirectoryId(cleartextDirectoryId: string): Promise<string> {
+    const dirHash = new TextEncoder().encode(cleartextDirectoryId);
+    // TODO https://github.com/encryption-alliance/unified-vault-format/pull/19 !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! MUST NEVER BE RELEASED LIKE THIS
```
Comment (chenkins) [inline frontend/src/common/crypto.ts:573]: @overheadhunter finalize

File: frontend/src/common/vaultconfig.ts [hunk 8]
```diff
@@ -35,12 +37,14 @@ export class VaultConfig {
 
     const vaultConfigToken = await vaultKeys.createVaultConfig(kid, hubConfig, jwtPayload);
     const rootDirHash = await vaultKeys.hashDirectoryId('');
-    return new VaultConfig(vaultConfigToken, rootDirHash);
+    return new VaultConfig(vaultConfigToken, vaultUvf, rootDirHash);
   }
 
   public async exportTemplate(): Promise<Blob> {
     const zip = new JSZip();
+    // TODO https://github.com/encryption-alliance/unified-vault-format/pull/19 what about vault.uvf? Not in template but only in vault dto?
```
Comment (chenkins) [inline frontend/src/common/vaultconfig.ts:45]: @overheadhunter what about vault.uvf? Not in template but only in vault dto?

Comment (overheadhunter) [reply frontend/src/common/vaultconfig.ts:45]: The vault will be either v8 or uvf, not both at the same time. When I said Hub will support both in parallel, I wasn't refering to the same vaults. New vault will use uvf, while keeping older ones is still supported with the existing mechanisms.

**Pull request merged**